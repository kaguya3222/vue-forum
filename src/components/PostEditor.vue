<template>
  <form @submit.prevent="addPost()">
    <div class="form-group">
      <textarea
        name=""
        id=""
        cols="30"
        rows="10"
        class="form-input"
        v-model="newPostText"
      ></textarea>
    </div>
    <div class="form-actions">
      <button class="btn-blue">Submit post</button>
    </div>
  </form>
</template>

<script>
import sourceData from "@/data.json";

export default {
  props: {
    threadId: {
      required: true,
      type: String
    }
  },
  data() {
    return {
      thread: sourceData.threads[this.threadId],
      newPostText: ""
    };
  },
  methods: {
    addPost() {
      const postId = "greatPost" + Math.random();
      const userId = "jUjmgCurRRdzayqbRMO7aTG9X1G2";
      sourceData.posts[postId] = {
        text: this.newPostText,
        publishedAt: Math.floor(Date.now() / 1000),
        threadId: this.threadId,
        userId,
        ".key": postId
      };
      sourceData.users[userId].posts[postId] = postId;
      this.$set(this.thread.posts, postId, postId);
      this.newPostText = "";
    }
  }
};
</script>
