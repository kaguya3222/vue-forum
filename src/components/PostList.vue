<template>
  <div class="post-list">
    <post-list-item v-for="post in posts" :key="post['.key']" :post="post" />
  </div>
</template>

<script>
import PostListItem from "./PostListItem";

export default {
  components: {
    "post-list-item": PostListItem
  },
  props: {
    posts: {
      required: true,
      type: Array
    }
  }
};
</script>

<style scoped></style>
