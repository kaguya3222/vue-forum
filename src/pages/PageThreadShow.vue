<template>
  <div class="col-large push-top">
    <div class="thread-header">
      <div class="thread-details">
        <h1>{{ thread.title }}</h1>
        <p>
          By <a href="#" class="link-unstyled">{{ user.name }}</a
          >, <app-date :unixDate="thread.publishedAt" />.
        </p>
      </div>
      <router-link
        :to="{ name: 'ThreadEdit', threadId: this.threadId }"
        class="btn-green btn-small"
        tag="button"
        >Edit</router-link
      >
    </div>

    <span class="hide-mobile text-faded text-small"
      >{{ repliesCount }} replies by {{ contributorsCount }} contributors</span
    >

    <post-list :posts="threadPosts" />

    <post-editor :threadId="threadId" />
  </div>
</template>

<script>
import PostList from "../components/PostList";
import PostEditor from "../components/PostEditor";
import { mapGetters } from "vuex";

export default {
  components: {
    "post-list": PostList,
    "post-editor": PostEditor
  },
  props: {
    threadId: {
      required: true,
      type: String
    }
  },
  computed: {
    ...mapGetters(["posts", "threads", "users", "countThreadReplies"]),
    contributorsCount() {
      const userIds = this.threadPosts.map(post => post.userId);
      return [...new Set(userIds)].length - 1;
    },
    threadPosts() {
      const postsIds = Object.values(this.thread.posts);
      return Object.values(this.posts).filter(post =>
        postsIds.includes(post[".key"])
      );
    },
    repliesCount() {
      return this.countThreadReplies(this.thread[".key"]);
    },
    thread() {
      return this.threads[this.threadId];
    },
    user() {
      return this.users[this.thread.userId];
    }
  }
};
</script>

<style scoped></style>
