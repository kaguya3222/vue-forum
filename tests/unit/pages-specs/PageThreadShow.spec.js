import { createLocalVue, shallowMount } from "@vue/test-utils";
import PageThreadShow from "../../../src/pages/PageThreadShow";
import PostList from "../../../src/components/PostList";
import PostEditor from "../../../src/components/PostEditor";
import mockedSourceData from "../mocks/mockedSourceData";
import Vuex from "vuex";
import sourceStore from "../../../src/store";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("PageThreadShow", () => {
  const store = new Vuex.Store(sourceStore);
  const wrapper = shallowMount(PageThreadShow, {
    propsData: {
      threadId: mockedSourceData.threads[0][".key"]
    },
    stubs: {
      "app-date": true
    },
    localVue,
    store
  });
  test("Shows thread data on thread page", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Accepts PostList as a child component", () => {
    expect(wrapper.contains(PostList)).toBe(true);
  });
  test("Accepts PostEditor as a child component", () => {
    expect(wrapper.contains(PostEditor)).toBe(true);
  });
  test("Creates posts array of certain thread", () => {
    const post = Object.values(mockedSourceData.posts)[0];
    const hasPost = wrapper.vm.threadPosts.some(
      threadPost => threadPost[".key"] === post[".key"]
    );
    expect(hasPost).toBe(true);
  });
});
