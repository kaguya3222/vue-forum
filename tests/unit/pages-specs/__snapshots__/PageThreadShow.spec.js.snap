// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageThreadShow Shows thread data on thread page 1`] = `
<div class="col-large push-top">
  <h1>What kind of pet do you have?</h1>
  <router-link-stub to="[object Object]" tag="button" class="btn-green btn-small">Edit</router-link-stub>
  <p>
    By <a href="#" class="link-unstyled">Robin</a>,
    <app-date-stub unixdate="1504035908"></app-date-stub>.
  </p>
  <post-list-stub posts="[object Object],[object Object],[object Object],[object Object]"></post-list-stub>
  <post-editor-stub threadid="-KsjWehQ--apjDBwSBCY"></post-editor-stub>
</div>
`;
